### PR TITLE
feat(compiler) : i18n explicit ID

### PR DESCRIPTION
**Pull request by mallex-actee**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
See [[i18n] enable control over translation message ID](https://github.com/angular/angular/issues/12941)

**What is the new behavior?**
Developer can now use specific ID in translations by adding a new "parameter" in the i18n attribute : 
```html
<p i18n="meaning|description@id">foo</p>
<!-- i18n: meaning1|desc1@@id1 -->bar<!-- /i18n -->
```

will produce for XMB : 

```xml
<msg id="id" meaning="meaning" description="description">foo</msg>
<msg id="id1" meaning="meaning1" description="description1">foo</msg>
```

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by maxime-allex** [comment]
@vicb I created new PR, because I'm not very good at rebasing. Now you have a clean PR.

File: modules/@angular/compiler/src/i18n/extractor_merger.ts [hunk 1]
```diff
@@ -330,15 +330,15 @@ class _Visitor implements html.Visitor {
   }
 
   // add a translatable message
-  private _addMessage(ast: html.Node[], meaningAndDesc?: string): i18n.Message {
+  private _addMessage(ast: html.Node[], i18nInfo?: string): i18n.Message {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/extractor_merger.ts]: what do you think or renaming `i18nInfo` to `msgInfo` / `msgMeta` ?

File: modules/@angular/compiler/src/i18n/extractor_merger.ts [hunk 2]
```diff
@@ -330,15 +330,15 @@ class _Visitor implements html.Visitor {
...
     if (ast.length == 0 ||
         ast.length == 1 && ast[0] instanceof html.Attribute && !(<html.Attribute>ast[0]).value) {
       // Do not create empty messages
       return;
     }
 
-    const [meaning, description] = _splitMeaningAndDesc(meaningAndDesc);
-    const message = this._createI18nMessage(ast, meaning, description);
+    const [meaning, description, id] = _parsei18nInfo(i18nInfo);
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/extractor_merger.ts]: `_parseMessageMeta` would look better, WDYT ?

File: modules/@angular/compiler/src/i18n/extractor_merger.ts [hunk 3]
```diff
@@ -496,8 +496,19 @@ function _getI18nAttr(p: html.Element): html.Attribute {
   return p.attrs.find(attr => attr.name === _I18N_ATTR) || null;
 }
 
-function _splitMeaningAndDesc(i18n: string): [string, string] {
-  if (!i18n) return ['', ''];
+function _parsei18nInfo(i18n: string): [string, string, string] {
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/extractor_merger.ts]: please return an obj now, would be clerarer

File: modules/@angular/compiler/src/i18n/extractor_merger.ts [hunk 4]
```diff
@@ -496,8 +496,19 @@ function _getI18nAttr(p: html.Element): html.Attribute {
...
+  if (!i18n) return ['', '', ''];
   const pipeIndex = i18n.indexOf('|');
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/extractor_merger.ts]: Please look for `@@` first.
- split left part on | for m|d,
- right part is id

add `const MEANING_SEPARATOR = "|"` and `const ID_SEPARATOR = "|"`

File: modules/@angular/compiler/test/i18n/digest_spec.ts [hunk 5]
```diff
@@ -6,10 +6,23 @@
...
   describe('digest', () => {
+    describe('digest', () => {
+      it('must return the ID if it\'s explicit', () => {
+        expect(digest({
+          nodes: [],
+          placeholders: {},
+          placeholderToMessage: {},
+          meaning: '',
+          description: '',
+          id: 'i'
```
Comment (vicb) [inline modules/@angular/compiler/test/i18n/digest_spec.ts]: add trailing "," here please - and maybe move id first (this is the important part)

**Comment by vicb** [comment]
Thanks !

I have added a last round of comments.

To rebase, update your master branch from upstream (or origin whatever it is name) and do a `git rebase master -i` in your feature branch.
That's pretty easy, do not hesitate to ask if you need help.

_Also please remove the WS before the colon in `feat(..) : ...`, thanks_

**Comment by vicb** [comment]
Also we discussed about an issue/PR on the doc repo, any progress on that front ?

File: modules/@angular/compiler/src/i18n/extractor_merger.ts (hunk 1 above)
Comment (maxime-allex) [inline modules/@angular/compiler/src/i18n/extractor_merger.ts]: @vicb OK for msgMeta.

**Comment by maxime-allex** [comment]
@vicb see the angular.io PR => [docs(compiler): add doc + examples for i18n explicit ID option](https://github.com/angular/angular.io/pull/2941)

File: modules/@angular/compiler/test/i18n/extractor_merger_spec.ts [hunk 6]
```diff
@@ -29,11 +32,45 @@ export function main() {
...
+        expect(
+            extract(
+                '<div i18n="d1@@i1"><span i18n-title="d2@@i2" title="single child">nested</span></div>'))
+            .toEqual([
+              [['<ph tag name="START_TAG_SPAN">nested</ph name="CLOSE_TAG_SPAN">'], '', 'd1', 'i1'],
+              [['single child'], '', 'd2', 'i2'],
+            ]);
+      });
+
+      it('should extract from attributes with ONLY id', () => {
```
Comment (vicb) [inline modules/@angular/compiler/test/i18n/extractor_merger_spec.ts]: nit: "id only"

**Comment by vicb** [comment]
Thanks, this PRs looks good. Only a couple more changes before this can be merged.

**Comment by vicb** [comment]
Thanks! it should get in tomorrow.

**Comment by kibbled** [comment]
Would this feature allow us to set IDs in our HTML/XLF files and avoid having to maintain IDs changing?

**Pull request closed without merging**